File: src/madpy/moleculargeometry.py
import re

from tequila import Molecule

from ._madpy_impl import MolecularGeometry as MolecularGeometryImpl
from .mrafunctionwrapper import MRAFunction3D


class MolecularGeometry:
    impl = None
    silent = False

    def __init__(self, geometry: str = None, units=None, silent=False, *args, **kwargs):
        self.silent = silent

        if units is None:
            if not self.silent:
                print(
                    "Warning: No units passed with geometry, assuming units are angstrom."
                )
            units = "angstrom"
        else:
            units = units.lower()
            if units in ["angstrom", "ang", "a", "å"]:
                units = "angstrom"
            elif units in ["bohr", "atomic units", "au", "a.u."]:
                units = "bohr"
            else:
                if not self.silent:
                    print(
                        "Warning: Units passed with geometry not recognized (available units are angstrom or bohr), assuming units are angstrom."
                    )
                units = "angstrom"

        self.impl = MolecularGeometryImpl(units)
        if geometry is not None:
            geometry = geometry.lower()
            geometry = geometry.strip()
            # Replace tabs with spaces
            geometry = geometry.replace("\t", " ")
            # Replace multiple whitespace characters with a single space
            re.sub(r"\s+", " ", geometry).strip()

            for line in geometry.split("\n"):
                data = line.split(" ")
                x = eval(data[1])
                y = eval(data[2])
                z = eval(data[3])
                s = data[0]
                self.add_atom(x, y, z, s)

    def check_units(self):
        return self.impl.units

    def add_atom(self, pos_x, pos_y, pos_z, symbol):
        self.impl.add_atom(pos_x, pos_y, pos_z, symbol)

    def to_json(self):
        return self.impl.to_json()

    def compute_nuclear_derivative(self, madworld, atom, axis):
        return MRAFunction3D(
            self.impl.compute_nuclear_derivative(madworld.impl, atom, axis)
        )

    def compute_second_nuclear_derivative(
        self, madworld, atom: int, axis1: int, axis2: int
    ):
        return MRAFunction3D(
            self.impl.compute_second_nuclear_derivative(
                madworld.impl, atom, axis1, axis2
            )
        )

    def nuclear_repulsion_derivative(self, atom: int, axis: int):
        return self.impl.nuclear_repulsion_derivative(atom, axis)

    def nuclear_repulsion_second_derivative(
        self, atom1: int, atom2: int, axis1: int, axis2: int
    ):
        return self.impl.nuclear_repulsion_second_derivative(atom1, atom2, axis1, axis2)

    def get_vnuc(self, madworld):
        return MRAFunction3D(self.impl.get_vnuc(madworld.impl))

    def get_nuclear_charge(self):
        return self.impl.get_nuclear_charge()

    def get_nuclear_repulsion(self):
        return self.impl.get_nuclear_repulsion()

    @property
    def n_electrons(self):
        return int(self.get_nuclear_charge())

    @property
    def n_core_electrons(self):
        return self.impl.get_core_n_electrons()

    # conversion from tequila molecule to molecular geometry
    def from_tq_mol(tq_mol, units="angstrom"):
        geometry = tq_mol.parameters.get_geometry_string(desired_units=units)
        return MolecularGeometry(geometry, units=units)

    # conversion from molecular geometry to tequila molecule
    def to_tq_mol(self, *args, **kwargs):
        pass

File: src/moleculargeometry.cpp
#include "moleculargeometry.hpp"

using namespace madness;

MolecularGeometry::MolecularGeometry(std::string units) : units(units) {
    if (units != "angstrom" && units != "bohr") {
        std::cout << "Warning: units of molecular geometry not recognized, assuming angstrom." << std::endl;
        this->units = "angstrom";
    }
    mol = madness::Molecule();
}

void MolecularGeometry::add_atom(double pos_x, double pos_y, double pos_z, std::string symbol) {
    double scale = 1e-10 / madness::constants::atomic_unit_of_length;
    if (units == "angstrom") {
        pos_x *= scale;
        pos_y *= scale;
        pos_z *= scale;
    }
    mol.add_atom(pos_x, pos_y, pos_z, (double)symbol_to_atomic_number(symbol), symbol_to_atomic_number(symbol));
}

std::string MolecularGeometry::to_json() {
    return mol.to_json().dump();
}

SavedFct<3> MolecularGeometry::molecular_potential_derivative(MadnessProcess<3>& mp, const int atom, const int axis) {
    double scale = 1e-10 / madness::constants::atomic_unit_of_length;
    madchem::MolecularDerivativeFunctor func(mol, atom, axis);
    real_function_3d op = real_factory_3d(*(mp.world)).functor(func).truncate_on_project().truncate_mode(0);
    if (units == "angstrom") {
        op = scale * op; // scale the function to Angstrom (units of the resulting function are hartree/angstrom)
    }
    return SavedFct<3>(op);
}

SavedFct<3> MolecularGeometry::molecular_potential_second_derivative(MadnessProcess<3>& mp, const int atom, const int axis1,
                                                                 const int axis2) {
    double scale = 1e-10 / madness::constants::atomic_unit_of_length;
    madchem::MolecularSecondDerivativeFunctor func(mol, atom, axis1, axis2);
    real_function_3d op = real_factory_3d(*(mp.world)).functor(func).truncate_on_project().truncate_mode(0);
    if (units == "angstrom") {
        op = scale * scale *
             op; // scale the function to Angstrom (units of the resulting function are hartree/angstrom**2)
    }
    return SavedFct<3>(op);
}

double MolecularGeometry::nuclear_repulsion_derivative(const int atom, const int axis) {
    double scale = 1e-10 / madness::constants::atomic_unit_of_length;
    double result = mol.nuclear_repulsion_derivative(atom, axis);
    if (units == "angstrom") {
        result = scale * result; // scale to Angstrom (units of the resulting function are hartree/angstrom)
    }
    return result;
}

double MolecularGeometry::nuclear_repulsion_second_derivative(const int atom1, const int atom2, const int axis1,
                                                              const int axis2) {
    double scale = 1e-10 / madness::constants::atomic_unit_of_length;
    double result = mol.nuclear_repulsion_second_derivative(atom1, atom2, axis1, axis2);
    if (units == "angstrom") {
        result = scale * scale * result; // scale to Angstrom (units of the resulting function are hartree/angstrom**2)
    }
    return result;
}

SavedFct<3> MolecularGeometry::get_vnuc(MadnessProcess<3>& mp) {
    PotentialManager pm = PotentialManager(mol, "");
    pm.make_nuclear_potential(*(mp.world));
    real_function_3d Vnuc = pm.vnuclear();
    return SavedFct<3>(Vnuc);
}